core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -0,0 +1,319 @@
+
+  type InitPidCallback = InitPidResult => Unit
+  type TxnMetadataUpdateCallback = Errors => Unit
+  type EndTxnCallback = Errors => Unit
+
+  /* Active flag of the coordinator */
+  private val isActive = new AtomicBoolean(false)
+
+  def handleInitPid(transactionalId: String,
+                    transactionTimeoutMs: Int,
+                    responseCallback: InitPidCallback): Unit = {
+    if (transactionalId == null || transactionalId.isEmpty) {
+      // if the transactional id is not specified, then always blindly accept the request
+      // and return a new pid from the pid manager
+      val pid = pidManager.nextPid()
+      responseCallback(InitPidResult(pid, epoch = 0, Errors.NONE))
+    } else if (!txnManager.isCoordinatorFor(transactionalId)) {
+      // check if it is the assigned coordinator for the transactional id
+      responseCallback(initTransactionError(Errors.NOT_COORDINATOR))
+    } else if (txnManager.isCoordinatorLoadingInProgress(transactionalId)) {
+      responseCallback(initTransactionError(Errors.COORDINATOR_LOAD_IN_PROGRESS))
+    } else if (!txnManager.validateTransactionTimeoutMs(transactionTimeoutMs)) {
+      // check transactionTimeoutMs is not larger than the broker configured maximum allowed value
+      responseCallback(initTransactionError(Errors.INVALID_TRANSACTION_TIMEOUT))
+    } else {
+      // only try to get a new pid and update the cache if the transactional id is unknown
+      txnManager.getTransaction(transactionalId) match {
+        case None =>
+          val pid = pidManager.nextPid()
+          val newMetadata: TransactionMetadata = new TransactionMetadata(pid = pid,
+            epoch = 0,
+            txnTimeoutMs = transactionTimeoutMs,
+            state = Empty,
+            topicPartitions = collection.mutable.Set.empty[TopicPartition],
+            timestamp = time.milliseconds())
+
+          val metadata = txnManager.addTransaction(transactionalId, newMetadata)
+
+          // there might be a concurrent thread that has just updated the mapping
+          // with the transactional id at the same time; in this case we will
+          // treat it as the metadata has existed and update it accordingly
+          metadata synchronized {
+            if (!metadata.equals(newMetadata))
+              metadata.epoch = (metadata.epoch + 1).toShort
+          }
+
+          responseCallback(initTransactionMetadata(metadata))
+
+        case Some(metadata) =>
+          metadata synchronized {
+            metadata.epoch = (metadata.epoch + 1).toShort
+          }
+          responseCallback(initTransactionMetadata(metadata))
+      }
+    }
+  }
+
+  private def validateTransactionalId(transactionalId: String): Errors =
+    if (transactionalId == null || transactionalId.isEmpty)
+      Errors.INVALID_REQUEST
+    else if (!txnManager.isCoordinatorFor(transactionalId))
+      Errors.NOT_COORDINATOR
+    else if (txnManager.isCoordinatorLoadingInProgress(transactionalId))
```

Comment:
@apurvam @junrao i already changed `handleTxnEmigration` to remove the inflight operations that we know about, i.e., the ones in purgatory. Once they complete they will error with `NOT_COORDINATOR`. However this is only useful for the case where we writing the Txn Markers.  
For the other cases, if the coordinator is still loading the cache and the `transactionalId` is not yet cached we will respond with `NOT_COORDINATOR`. However, if the cache is  is loading  it may have an old record for the `transactionalId`. Should we do a check in the callback, i.e.,
```
if (!isCoordinatorLoadingInProgress(transactionalId))
          responseError = Errors.NOT_COORDINATOR
else
 // complete
````